tft.py

```diff
@@ -0,0 +1,189 @@
+                f.write(template)
+
+        for language in font_files:
+            use = template
+            code = os.path.basename(language)[11:-4]
+
+            with open(language, "r", encoding="utf-8") as f:
+                for line in f:
+                    if line.startswith("tr"):
+                        key, val = line.split("=", 1)
+                        key = key[4:-2]
+                        val = val[2:-2]
+                        use = use.replace(key, val)
+
+            with open(os.path.join(output, code + ".json"), "w", encoding="utf-8") as f:
+                f.write(use)
+
+    def parse_items(self, map22):
+        item_collection = [x for x in map22.entries if x.type == "TftItemData"]
+        items = []
+
+        for item in item_collection:
+            name = item.getv("mName")
+            if "Template" in name or name == "TFT_Item_Null":
+                continue
+
+            effects = []
+            for effect in item.getv("effectAmounts") if item.get("effectAmounts") else []:
+                effects.append({"name": effect.getv("name"), "value": effect.getv("value")})
+
+            items.append(
+                {
+                    "internal": {"hex": item.path.hex(), "name": item.getv("mName")},
+                    "id": item.getv("mId"),
+                    "name": item.getrv("c3143d66"),
+                    "desc": item.getrv("765f18da"),
+                    "icon": item.getv("mIconPath"),
+                    "from": [x.hex() for x in item.getrv("8b83ba8a")] if item.getr("8b83ba8a") else [],
+                    "effects": effects,
+                }
+            )
+
+        # replace the hex key gotten earlier with the id
+        for item in items:
+            # there must be some list comprehension way of doing this...
+            for old in item["from"]:
+                key = [x for x in items if x["internal"]["hex"] == old][0]["id"]
+                item["from"].insert(0, key)
+                item["from"].remove(old)
+
+        return items
+
+    def parse_champs(self, map22, traits, character_folder):
+        champ_collection = [x for x in map22.entries if x.type == "TftShopData"]
+        champs = []
+
+        for champ in champ_collection:
+            name = champ.getv("mName")
+
+            if not name.startswith("TFT_") or name == "TFT_Template":
+                continue
+
+            char_bin = BinFile(os.path.join(character_folder, name[4:], name[4:] + ".bin"))
+            champ_id = [x.getv("characterToolData").getv("championId") for x in char_bin.entries if x.type == "CharacterRecord"][0]
+
+            champ_traits = []
+            tft_bin = BinFile(os.path.join(character_folder, name, name + ".bin"))
+            record = [x for x in tft_bin.entries if x.type == "TFTCharacterRecord"][0]
+            for trait in record.getv("mLinkedTraits"):
+                champ_traits.append([x["internal"]["name"] for x in traits if x["internal"]["hash"] == trait.hex()][0])
```

Comment:
Similarly to items, you can convert `traits` to a hash-to-name dict first (at the beginning of the method) and then use regular indexing.
```py
trait_names = {x["internal"]["hash"]: x["internal"]["name"] for x in traits}
...
champ_traits = [trait_names[trait.h] for trait in record["mLinkedTraits"].value]